### PR TITLE
Add root expression profiler

### PR DESCRIPTION
**Pull request by azimafroozeh**
#1495.

### CODE REVIEW

**Comment by azimafroozeh** [comment]
@peterboncz

**Comment by hannesmuehleisen** [comment]
We are currently introducing default thread sanitisation in #1639, which found a bunch of bugs related to profiling. Maybe its best to wait till that is merged.

File: src/function/table/pragma_detailed_profiling_output.cpp
Comment (Mytherin) [inline src/function/table/pragma_detailed_profiling_output.cpp:42]: Is it time or is it cycles per tuple? If it depends it is better to have two columns (`TIME` and `CYCLES_PER_TUPLE`)  where either time or cycles_per_tuple is set to NULL when not available.

File: third_party/concurrentqueue/concurrentqueue.h
Comment (Mytherin) [inline third_party/concurrentqueue/concurrentqueue.h:1464]: Accidental change?

File: src/function/table/pragma_detailed_profiling_output.cpp
Comment (Mytherin) [inline src/function/table/pragma_detailed_profiling_output.cpp:130]: What exactly does this computation do? Can we do this as a separate step instead of as a big ternary?

File: src/function/table/pragma_detailed_profiling_output.cpp
Comment (Mytherin) [inline src/function/table/pragma_detailed_profiling_output.cpp:81]: Same here, can this be done not as a big ternary but as a clear separate step that explains what's going on?

File: src/function/table/pragma_detailed_profiling_output.cpp
Comment (Mytherin) [inline src/function/table/pragma_detailed_profiling_output.cpp:82]: This needs either an assertion or a check that sample_tuples_count != 0

File: src/function/table/pragma_detailed_profiling_output.cpp
Comment (Mytherin) [inline src/function/table/pragma_detailed_profiling_output.cpp:130]: This needs either an assertion or a check that et->sample_tuples_count != 0

File: src/include/duckdb/common/cycle_counter.hpp
Comment (Mytherin) [inline src/include/duckdb/common/cycle_counter.hpp:37]: For readability, please take the 50 and 100 out and put them in some static constexpr's (called e.g. sampling rate, sampling variance)

File: src/include/duckdb/common/cycle_counter.hpp
Comment (Mytherin) [inline src/include/duckdb/common/cycle_counter.hpp:33]: This code block seems like a repeat of ProcessSample(). Can this function just call ProcessSample()?

File: src/include/duckdb/common/cycle_counter.hpp
Comment (Mytherin) [inline src/include/duckdb/common/cycle_counter.hpp:25]: Perhaps call these functions BeginSample and EndSample, to make it more obvious that they need to be called together and in that order.

File: src/include/duckdb/common/cycle_counter.hpp
Comment (Mytherin) [inline src/include/duckdb/common/cycle_counter.hpp:22]: The random engine always uses the same seed. Is that intended?

File: src/include/duckdb/common/cycle_counter.hpp
Comment (Mytherin) [inline src/include/duckdb/common/cycle_counter.hpp:66]: Does all of this need to be public? Can the private tag come back?

File: src/include/duckdb/function/table/sqlite_functions.hpp
Comment (Mytherin) [inline src/include/duckdb/function/table/sqlite_functions.hpp:35]: Unused?

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! This looks better already. Some initial comments about the code.

It also appears there are some memory leaks on the release assertions CI run. Could you have a look there? 

As @hannesmuehleisen mentioned we are introducing the thread sanitizer, and we had found some problems relating to profiling and multi-threading that we fixed there. Could you perhaps enable the detailed profiling in the multi-threading tests (e.g. the tests in parallelism/interquery and parallelism/intraquery)? You can test these locally with the thread sanitizer enabled by compiling with `THREADSAN=1 make` or enabling the cmake option `ENABLE_THREAD_SANITIZER`. Once that works and #1639 is merged the tests should pass there as well.

File: src/function/table/pragma_detailed_profiling_output.cpp
Comment (azimafroozeh) [reply src/function/table/pragma_detailed_profiling_output.cpp:42]: It depends on the BUILD_RDTSC flag. If it is not enabled, the cycle counter measures time.

**Comment by azimafroozeh** [comment]
> It also appears there are some memory leaks on the release assertions CI run. Could you have a look there?

On the same [ci run](https://github.com/azimafroozeh/duckdb/actions/runs/816581117)! on my repo, there was no memory leak. 
 Also, I could not reproduce this on intel MacBook, m1 MacBook, and also the president machine. Any suggestions?

> As @hannesmuehleisen mentioned we are introducing the thread sanitizer, and we had found some problems relating to profiling and multi-threading that we fixed there. Could you perhaps enable the detailed profiling in the multi-threading tests (e.g. the tests in parallelism/interquery and parallelism/intraquery)? You can test these locally with the thread sanitizer enabled by compiling with `THREADSAN=1 make` or enabling the cmake option `ENABLE_THREAD_SANITIZER`. Once that works and #1639 is merged the tests should pass there as well.

Sure. I think threads flush down information to the main profiler without any locking mechanism which is problematic. Also in the main executor, for every data chunk, the flushing and finalizingoperatorstate functions will be executed which is not optimal.

File: src/function/table/pragma_detailed_profiling_output.cpp
Comment (Mytherin) [reply src/function/table/pragma_detailed_profiling_output.cpp:42]: Those should be two separate columns. Mixing two different data points in a single column is asking for trouble.

File: src/function/table/pragma_detailed_profiling_output.cpp
Comment (azimafroozeh) [reply src/function/table/pragma_detailed_profiling_output.cpp:130]: Yes, just to make sure that the number of cycles is a valid number.

**Comment by Mytherin** [comment]
Have you merged with master? The CI run on the PR tests against a merge with master, the CI run on your repo tests only your branch.

I would try running exactly the code as specified in the main.yml file for release assertions. Otherwise, the CI run also contains a stack trace, perhaps that can be used to figure out where the problem comes from.

File: src/include/duckdb/common/cycle_counter.hpp
Comment (azimafroozeh) [reply src/include/duckdb/common/cycle_counter.hpp:33]: This is like an optimization. because for a function inside an expression, we don't need to count the number of process tuples. it would happen already in the expression-root cycle counter.

Comment (Mytherin) [reply src/include/duckdb/common/cycle_counter.hpp:33]: That's fine, but that doesn't require the code to be copy pasted, no?

Also care would need to be taken in that case, since tuple_count will be zero on the expression counter, right? There are several places where we divide by the tuple_count, which could lead to a division by zero.

Comment (azimafroozeh) [reply src/include/duckdb/common/cycle_counter.hpp:33]: The tuple_count will be zero, but it will be never used. The tuple_count from Expressionexecutorstate (expression_root) will be used in those cases. Potentially the function cycle-counter and root-expression cycle-counter could be divided into two classes which one inherit from the other one. This way, we also save space.

File: src/include/duckdb/common/cycle_counter.hpp
Comment (Mytherin) [reply src/include/duckdb/common/cycle_counter.hpp:22]: This seems like a premature optimization with potential of causing issues. How does delegating this tuple counter interact with e.g. CASE expressions or conjunctions, that can cause the execution of entire branches of expressions to be skipped?

Comment (azimafroozeh) [reply src/include/duckdb/common/cycle_counter.hpp:22]: Hmm, fair. I have disabled this optimization.

**Comment by Mytherin** [comment]
Thanks for the changes! Could you enable the detailed profiling in the multi-threading tests (e.g. the tests in parallelism/interquery and parallelism/intraquery), and re-enable the previously broken test `test/sql/pragma/test_pragma_output.test` (by removing the `mode skip` at the beginning of the test)?

File: test/sql/parallelism/interquery/test_concurrentupdate.cpp
Comment (Mytherin) [inline test/sql/parallelism/interquery/test_concurrentupdate.cpp:25]: __TEST_DIR__ like this will only work in the sqllogictests. Do you need to write the profiling output to a file? If so, use TestCreatePath from test_helpers.hpp.

Comment (azimafroozeh) [reply test/sql/parallelism/interquery/test_concurrentupdate.cpp:25]: I think if the profiling output doesn’t go into a file, it would spam the testing output with the result of each query.

File: src/include/duckdb/main/query_profiler.hpp
Comment (Mytherin) [inline src/include/duckdb/main/query_profiler.hpp:137]: A public lock is usually asking for trouble. Can this be made private, and obtained by e.g. the Flush method?

**Review by Mytherin — COMMENTED** [review]
Thanks for making the changes! Some more requests (and some nitpicking about code style :)):

File: src/common/tree_renderer.cpp
Comment (azimafroozeh) [reply src/common/tree_renderer.cpp]: Ahhh, super thanks for this link.

**Comment by Mytherin** [comment]
This looks fine from my side now. @lnkuiper will do another pass over the code, after which I think we can merge it.

File: src/execution/operator/aggregate/physical_simple_aggregate.cpp
Comment (lnkuiper) [inline src/execution/operator/aggregate/physical_simple_aggregate.cpp:213]: Why does `FinalizeOperatorState` need to be overridden here? Do you need the function of the base class not to be called?

File: src/main/client_context.cpp
Comment (lnkuiper) [inline src/main/client_context.cpp:105]: The function `query_profiler_history->GetPrevProfilers()` is being called many times here in the span of a  few lines. This would look cleaner if you do this once:
```
auto &prev_profilers = query_profiler_history->GetPrevProfilers();
```

And calling e.g. `size()` and `pop_front()` on the reference.

**Review by lnkuiper — CHANGES_REQUESTED** [review]
Looking good overall just some minor nitpickings. Looking forward to performance improvement opportunities being revealed as a result of your work!

File: src/execution/operator/aggregate/physical_simple_aggregate.cpp
Comment (azimafroozeh) [reply src/execution/operator/aggregate/physical_simple_aggregate.cpp:213]: For this PR, FinalizeOperatorState does not need to be overridden. But there is an aggregate function in this operator that I think needs to be monitored. Therefore FinalizeOperatorState needs to be overridden to flush down the monitored information.

File: src/execution/expression_executor.cpp
Comment (azimafroozeh) [reply src/execution/expression_executor.cpp:29]: It could be moved to ```Initialize``` and being cleaner. But the ```Initialize``` function initializes the ExpressionExecutor object, not the ExpressionExecutorState. It could become buggy since, after any ExpressionExecutorState initiation, there should be an ExpressionExecutor and get initialized. Off-course, the ExpressionExecutorState is only used here.

**Review by lnkuiper — APPROVED** [review]
All good from my side now

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**
